design/velero-upgrade.md

```diff
@@ -0,0 +1,85 @@
+
+The manual approach works for the existing scenarios, however, if we want to do some complex tasks during upgrade (i.e., installing a new module or changing/adding complex parameters), it will be very unfriendly in the user experiences.  
+For an concrete example, in v1.10, we have done some refactors to Velero's pod volume backup (a.k.a Restic backup formerly), two of the refactors are:  
+- Rename some Velero server and daemonset parameters
+- Rename Velero daemonset from "restic" to "node-agent"
+
+As a result, the existing upgrade approach is not sufficient:
+- If we still use manual steps, we have to tell users the detailed parameters that have been renamed, and users then needs to check their existence and modify them one by one.
+- If we still use manual steps, users have to delete the old restic daemonset and run ```velero install``` with the full parameters they specified during the fresh installation, because for one thing, Kubernetes doesn't support to rename a resource; for another, Velero doesn't provide a way to install a single resource only. 
+
+## Solution
+
+We want to add a new command besides ```velero install``` called ```velero upgrade```. In a upgrade case, users could call ```velero upgrade``` without bothering to know the details. This bring below benefits:  
+- Solve the upgrade problems in v1.10
+- Make Velero more user friendly, because the upgrade process is significantly simplified
+- Support future evolvements. For example, in future, Velero will have various data movement features, for which we will add new moudles inevitably. Then we can add enhancement incrementally to ```velero upgrade```, because inside the command, we can do whatever complex things 
+
+The ```velero upgrade``` command will fully replace the existing manual steps in the aforementioned Velero document.
+
+## Design
+
+At present, ```velero upgrade``` will do below things:
+- Get the new version from the client binary itself, that is, the upgrade target version
+- Get the old version from the existing velero server, that is, the current version
+- Retrieve Velero server deployment object
+- Retrieve Velero daemonset object
+- Iterate from current version through target version, apply the changes (i.e., the image, name and parameter changes) to the deployment object once per version. For example, when upgrading from 1.9 to 1.11, we need to apply the changes between 1.9 and 1.10 first, then apply the changes between 1.10 and 1.11. See the details below for how to drive the changes apply
+- Do the same changes apply to Velero daemonset object as the deployment object
+- Delete the existing Velero deployment
+- Delete the existing Velero daemonset
+- Extract the new CRDs from the client binary and apply them
+- Create the new Velero daemonset from the modified daemonset object
+- Create the new Velero server deployment from the modified deployment object
```

Comment:
@Lyndon-Li Once the new Velero deployment is created and is in running state, should we also check up on the state of prior BSLs that were already configured, whether they have available status or not post upgrade completion ?

Reply:
This is related to this [discussion](https://github.com/vmware-tanzu/velero/pull/5445#discussion_r996663794) , that is, whether we update the CRs during `velero upgrade`:

- IMO, we don't need to update CRs, we should use kubernetes built in approaches to manage CRs/CR versions. If so, since we don't change CRs, we don't need to verify their status after upgrade
- If we do need to update CRs, then we need to verify the status for the updated CRs

Let's wait and see the result of the [discussion](https://github.com/vmware-tanzu/velero/pull/5445#discussion_r996663794) and then decide the verification accordingly.